### PR TITLE
Enable String parameters in functions, and other small refinements.

### PR DESCRIPTION
**Pull request by schungx**
Try out with a small PR...

### CODE REVIEW

**Comment by schungx** [comment]
@jhwgh1968   just trying a small PR.  Looks like it works fine at this point...

I deeply suspect that Visual Studio probably does something weird with rebasing...

**Comment by schungx** [comment]
@jhwgh1968   I added support for `String` parameters in functions (before it won't match `ImmutableString` with `String`, only with `&str`).

**Comment by jhwgh1968** [comment]
It looks like the rebase worked, but you rebased onto a old version of the branch.

I am reviewing a fixed version on my own PC.

File: codegen/src/function.rs [hunk 1]
```diff
@@ -611,15 +655,24 @@ impl ExportedFn {
...
                                                    .clone().cast::<ImmutableString>())
                             }
                             _ => panic!("internal error: why wasn't this found earlier!?"),
                         },
+                        &syn::Type::Path(ref p) if p.path == string_type_path => {
+                            is_string = true;
+                            is_ref = false;
+                            quote_spanned!(arg_type.span()=>
+                                                   mem::take(args[#i])
+                                                   .clone().cast::<#arg_type>())
```
Comment (jhwgh1968) [inline codegen/src/function.rs:671]: You appear to have created a `Dynamic::take_immutable_string` method for just this purpose. Would using that be cheaper than a `clone()`?

**Comment by jhwgh1968** [comment]
Just one comment. Once you fix it, you can try a rebase to update it, or you can ask me to fix it.

File: codegen/src/function.rs (hunk 1 above)
Comment (schungx) [reply codegen/src/function.rs:671]: Yeah, good idea!

**Comment by schungx** [comment]
Oops.  Accidentally merged it into `plugins`...

**Comment by jhwgh1968** [comment]
Looking at the diff, a number of things were actually reversed, including `doc` updates and the `Position` argument changes I made to `call`. I will open a new PR to fix those.

In addition, there were a bunch of rustfmt changes which got reverted. I will solve that by opening a separate PR to enforce clean rustfmt before merge.

**Comment by schungx** [comment]
> Looking at the diff, a number of things were actually reversed, including `doc` updates and the `Position` argument changes I made to `call`. I will open a new PR to fix those.

Yeah, Git still keeps on doing weird things that I don't understand...

**Comment by jhwgh1968** [comment]
As noted on #38, I was mistaken about this, and reading the diff wrong. The merge commit, in the end, was okay.

**Pull request merged**